### PR TITLE
add configuration content compare when someone want to publish

### PR DESCRIPTION
**Pull request by svnshing**
## What's the purpose of this PR

add profile compare when someone want to publish,particularly convenient for the JSON string

## Which issue(s) this PR fixes:
Fixes #

## Brief changelog

do add compare when someone want to publish the profile,for json particularly convenient
K20200310-170432

Follow this checklist to help us incorporate your contribution quickly and easily:

- [ ] Read the [Contributing Guide](https://github.com/ctripcorp/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit tests to verify the code.
- [ ] Run `mvn clean test` to make sure this pull request doesn't break anything.

### CODE REVIEW

File: apollo-portal/src/main/resources/static/scripts/directive/release-modal-directive.js [hunk 1]
```diff
@@ -59,23 +59,23 @@ function releaseModalDirective($translate, toastr, AppUtil, EventManager, Releas
                     scope.toReleaseNamespace.releaseTitle,
                     scope.releaseComment,
                     scope.isEmergencyPublish).then(
-                        function (result) {
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/scripts/directive/release-modal-directive.js]: please keep the origin format to avoid unnecessary changes

File: apollo-portal/src/main/resources/static/scripts/directive/release-modal-directive.js [hunk 2]
```diff
@@ -88,38 +88,38 @@ function releaseModalDirective($translate, toastr, AppUtil, EventManager, Releas
...
-                            scope.toReleaseNamespace.lockOwner = undefined;
-
-                            //check rules
-                            if (!scope.toReleaseNamespace.rules
-                                || !scope.toReleaseNamespace.rules.ruleItems
-                                || !scope.toReleaseNamespace.rules.ruleItems.length) {
-
-                                scope.toReleaseNamespace.viewType = 'rule';
-                                AppUtil.showModal('#grayReleaseWithoutRulesTips');
+                    function (result) {
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/scripts/directive/release-modal-directive.js]: please keep the origin format to avoid unnecessary changes

File: apollo-portal/src/main/resources/static/views/component/release-modal.html [hunk 3]
```diff
@@ -51,103 +51,133 @@ <h4 class="modal-title" ng-show="toReleaseNamespace.isBranch && toReleaseNamespa
...
-                                        {{'Component.Publish.NoPublishedValue' | translate }}
-                                    </th>
-                                    <th>
-                                        {{'Component.Publish.ModifyUser' | translate }}
-                                    </th>
-                                    <th>
-                                        {{'Component.Publish.ModifyTime' | translate }}
-                                    </th>
-                                </tr>
+                            <tr>
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/views/component/release-modal.html]: please keep the origin format to avoid unnecessary changes

File: apollo-portal/src/main/resources/static/views/component/release-modal.html [hunk 4]
```diff
@@ -51,103 +51,133 @@ <h4 class="modal-title" ng-show="toReleaseNamespace.isBranch && toReleaseNamespa
...
-                                    <td width="25%" title="{{config.newValue}}">
-                                        <span ng-bind="config.newValue"></span>
-                                    </td>
-                                    <td width="15%" ng-bind="config.item.dataChangeLastModifiedBy">
-                                    </td>
-                                    <td width="15%"
-                                        ng-bind="config.item.dataChangeLastModifiedTime | date: 'yyyy-MM-dd HH:mm:ss'">
-                                    </td>
-                                </tr>
+                            <tr ng-repeat="config in toReleaseNamespace.items"
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/views/component/release-modal.html]: please keep the origin format to avoid unnecessary changes

File: apollo-portal/src/main/resources/static/views/component/release-modal.html [hunk 5]
```diff
@@ -51,103 +51,133 @@ <h4 class="modal-title" ng-show="toReleaseNamespace.isBranch && toReleaseNamespa
...
-                                        {{'Component.Publish.GrayNoPublishedValue' | translate }}
-                                    </th>
-                                    <th>
-                                        {{'Component.Publish.ModifyUser' | translate }}
-                                    </th>
-                                    <th>
-                                        {{'Component.Publish.ModifyTime' | translate }}
-                                    </th>
-                                </tr>
+                            <tr>
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/views/component/release-modal.html]: please keep the origin format to avoid unnecessary changes

File: apollo-portal/src/main/resources/static/views/component/release-modal.html [hunk 6]
```diff
@@ -51,103 +51,133 @@ <h4 class="modal-title" ng-show="toReleaseNamespace.isBranch && toReleaseNamespa
...
-                                    <td width="20%" title="{{config.newValue}}">
-                                        <span ng-bind="config.newValue"></span>
-                                    </td>
-                                    <td width="10%" ng-bind="config.item.dataChangeLastModifiedBy">
-                                    </td>
-                                    <td width="15%"
-                                        ng-bind="config.item.dataChangeLastModifiedTime | date: 'yyyy-MM-dd HH:mm:ss'">
-                                    </td>
-                                </tr>
+                            <tr ng-repeat="config in toReleaseNamespace.branchItems"
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/views/component/release-modal.html]: please keep the origin format to avoid unnecessary changes

File: apollo-portal/src/main/resources/static/views/component/release-modal.html [hunk 7]
```diff
@@ -51,103 +51,133 @@ <h4 class="modal-title" ng-show="toReleaseNamespace.isBranch && toReleaseNamespa
...
+                                </td>
+                                <td width="10%" ng-bind="config.item.dataChangeLastModifiedBy">
+                                </td>
+                                <td width="10%"
+                                    ng-bind="config.item.dataChangeLastModifiedTime | date: 'yyyy-MM-dd HH:mm:ss'">
+                                </td>
+                                <td width="5%">
+                                    <span>
+                                        <button type="button" class="btn btn-success"
+                                                ng-click="compare(config,0)">{{'Component.Publish.Text.Compare' | translate
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/views/component/release-modal.html]: Is it possible to only have one compare button and the program parse the actual format? 
I don't think it's extensible if we need to add xml compare, yml compare in the future.

image

File: apollo-portal/src/main/resources/static/scripts/directive/release-modal-directive.js (hunk 2 above)
Comment (svnshing) [reply apollo-portal/src/main/resources/static/scripts/directive/release-modal-directive.js]: ok,I will do it.Thank you

File: apollo-portal/src/main/resources/static/scripts/directive/release-modal-directive.js (hunk 1 above)
Comment (svnshing) [reply apollo-portal/src/main/resources/static/scripts/directive/release-modal-directive.js]: ok,I will do it.Thank you

File: apollo-portal/src/main/resources/static/views/component/release-modal.html (hunk 3 above)
Comment (svnshing) [reply apollo-portal/src/main/resources/static/views/component/release-modal.html]: ok,I will do it.Thank you

File: apollo-portal/src/main/resources/static/views/component/release-modal.html (hunk 4 above)
Comment (svnshing) [reply apollo-portal/src/main/resources/static/views/component/release-modal.html]: ok,I will do it.Thank you

File: apollo-portal/src/main/resources/static/views/component/release-modal.html (hunk 5 above)
Comment (svnshing) [reply apollo-portal/src/main/resources/static/views/component/release-modal.html]: ok,I will do it.Thank you

File: apollo-portal/src/main/resources/static/views/component/release-modal.html (hunk 7 above)
Comment (svnshing) [reply apollo-portal/src/main/resources/static/views/component/release-modal.html]: what you say is correct,maybe because I use json more.

Comment (svnshing) [reply apollo-portal/src/main/resources/static/views/component/release-modal.html]: I think Xml compare can use text compare model,maybe most people will use json cofiguartion in the text for applocation.For your word,maybe you want to support the xml comapre or other format,and you have a better position to place this configuration function?

Comment (svnshing) [reply apollo-portal/src/main/resources/static/views/component/release-modal.html]: also can support all format,I will improve this after,and merge request this after I finish improved.Thank you very much.

Comment (nobodyiam) [reply apollo-portal/src/main/resources/static/views/component/release-modal.html]: What I am proposing is to only show one button to compare, which is friendly to the user. Then for json format, we could use json compare mode, and for other formats, maybe we could just use text compare mode.

**Comment by nobodyiam** [comment]
@svnshing are you still working on this?
